cloudevents-sdk-warp/src/filter.rs

```diff
@@ -0,0 +1,136 @@
+use crate::server_request::request_to_event;
+
+use cloudevents::Event;
+use warp::http::HeaderMap;
+use warp::Filter;
+use warp::Rejection;
+
+#[derive(Debug)]
+pub struct EventFilterError {
+    message: String,
+}
+
+impl warp::reject::Reject for EventFilterError {}
+
+///
+/// # Extracts [`cloudevents::Event`] event from incoming request
+///
+/// ```
+/// use cloudevents_sdk_warp::filter::to_event;
+/// use warp::Filter;
+/// use warp::Reply;
+///
+/// let routes = warp::any()
+///    .and(to_event())
+///    .map(|event| {
+///         // do something with the event
+///     }
+///     );
+/// ```
+///
+pub fn to_event() -> impl Filter<Extract = (Event,), Error = Rejection> + Copy {
+    warp::header::headers_cloned()
+        .and(warp::body::bytes())
+        .and_then(create_event)
+}
+
+async fn create_event(headers: HeaderMap, body: bytes::Bytes) -> Result<Event, Rejection> {
+    request_to_event(headers, body).map_err(|e| {
+        warp::reject::custom(EventFilterError {
+            message: e.to_string(),
+        })
+    })
+}
+
+#[cfg(test)]
+mod tests {
+    use super::to_event;
+    use url::Url;
+    use warp::test;
+
+    use chrono::Utc;
+    use cloudevents::{EventBuilder, EventBuilderV10};
+    use serde_json::json;
+    use std::str::FromStr;
+
+    #[tokio::test]
+    async fn test_request() {
+        let time = Utc::now();
+        let expected = EventBuilderV10::new()
+            .id("0001")
+            .ty("example.test")
+            .source("http://localhost/")
+            .time(time)
+            .extension("someint", "10")
+            .build()
+            .unwrap();
+
+        let result = test::request()
+            .method("POST")
+            .header("ce-specversion", "1.0")
+            .header("ce-id", "0001")
+            .header("ce-type", "example.test")
+            .header("ce-source", "http://localhost/")
+            .header("ce-someint", "10")
+            .header("ce-time", time.to_rfc3339())
+            .filter(&to_event())
+            .await
+            .unwrap();
+
+        assert_eq!(expected, result);
+    }
+
+    #[tokio::test]
+    async fn test_bad_request() {
+        let time = Utc::now();
+
+        let result = test::request()
+            .method("POST")
+            .header("ce-specversion", "BAD SPECIFICATION")
+            .header("ce-id", "0001")
+            .header("ce-type", "example.test")
+            .header("ce-source", "http://localhost/")
+            .header("ce-someint", "10")
+            .header("ce-time", time.to_rfc3339())
+            .filter(&to_event())
+            .await;
+
+        assert!(result.is_err());
+        let rejection = result.unwrap_err();
+
+        let reason = rejection.find::<super::EventFilterError>().unwrap();
```

Comment:
400 Bad request check? Is there a way to do it?

Reply:
i haven't found a way. Rejections are a bit strange in warp, user should provide `recover` method and handle error

Reply:
Weird but ok :)

Reply:
In any case, the error struct is public, so user can in the recover handler match the proper error and return 400